### PR TITLE
8323243: JNI invocation of an abstract instance method corrupts the stack

### PR DESCRIPTION
**Pull request by dholmes-ora**
See JBS for details.

TL;DR - if an instance method is abstract then JNI front-end should throw AbstractMethodError.

Testing:
  - new regression test
  -  tiers 1-3 (sanity)

Thanks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323243](https://bugs.openjdk.org/browse/JDK-8323243): JNI invocation of an abstract instance method corrupts the stack (**Bug** - P4)

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [284d0e1b](https://git.openjdk.org/jdk/pull/17337/files/284d0e1b8dd14467f3e2a27910166579ec6d8111)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17337/head:pull/17337` \
`$ git checkout pull/17337`

Update a local copy of the PR: \
`$ git checkout pull/17337` \
`$ git pull https://git.openjdk.org/jdk.git pull/17337/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17337`

View PR using the GUI difftool: \
`$ git pr show -t 17337`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17337.diff">https://git.openjdk.org/jdk/pull/17337.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17337#issuecomment-1884080716)

### CODE REVIEW

File: test/hotspot/jtreg/runtime/jni/abstractMethod/AbstractMethod.jasm [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+       stack 1 locals 1
+    {
+        aload_0;
+        invokespecial Method java/lang/Object."<init>":"()V";
+        return;
+    }
+
+    public abstract Method "abstractM":"()V";
+
+}
```
Comment (coleenp) [inline test/hotspot/jtreg/runtime/jni/abstractMethod/AbstractMethod.jasm:43]: some newline missing.

File: test/hotspot/jtreg/runtime/jni/abstractMethod/TestJNIAbstractMethod.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    // Invokes an abstract method from JNI and throws AbstractMethodError.
+    private static native void invokeAbstractM(Class<?> AMclass, AbstractMethod receiver);
+
+    static {
+        System.loadLibrary("JNIAbstractMethod");
+    }
+
+    public static void main(String[] args) {
+        AbstractMethod obj = new AbstractMethod();
```
Comment (coleenp) [inline test/hotspot/jtreg/runtime/jni/abstractMethod/TestJNIAbstractMethod.java:51]: So this doesn't result in an exception when you new this class?  Can you call this AbstractMethodClass or something like that?

**Review by coleenp — APPROVED** [review]
I have a couple of minor comments but looks good.

File: test/hotspot/jtreg/runtime/jni/abstractMethod/TestJNIAbstractMethod.java (hunk 2 above)
Comment (shipilev) [reply test/hotspot/jtreg/runtime/jni/abstractMethod/TestJNIAbstractMethod.java:51]: It would also be good to have a positive test here: calling the method from plain Java should throw AME as well.

**Review by shipilev — APPROVED** [review]
Ouch. Looks good, with minor test suggestion.

File: test/hotspot/jtreg/runtime/jni/abstractMethod/AbstractMethod.jasm (hunk 1 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/jni/abstractMethod/AbstractMethod.jasm:43]: Fixed

File: test/hotspot/jtreg/runtime/jni/abstractMethod/TestJNIAbstractMethod.java (hunk 2 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/jni/abstractMethod/TestJNIAbstractMethod.java:51]: The class itself is not marked abstract so you can create instances of it. The VM spec does not require that a non-abstract class have no abstract methods.

I have renamed to `AbstractMethodClass`.

**Comment by dholmes-ora** [comment]
Thanks for the reviews @coleenp  and @shipilev !

@shipilev I don't see your "minor test suggestion". ??

File: test/hotspot/jtreg/runtime/jni/abstractMethod/TestJNIAbstractMethod.java (hunk 2 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/jni/abstractMethod/TestJNIAbstractMethod.java]: > It would also be good to have a positive test here: calling the method from plain Java should throw AME as well.

Finally found this comment! No idea where it was hiding.

Okay I've added a direct Java test too (not sure I'd call it a "positive test" though : ) ). I actually thought it may not compile but it seems compile time method resolution doesn't care about abstract (which I can understand for interface invocations but not classes).

**Review by shipilev — APPROVED** [review]
This looks good.

Minor nit still: I think we gotta try the Java side first, because that checks the expected behavior. Right now, if JNI fails, we don't really know if that is a product bug (= "JNI failed to throw AME when it should"), or a test bug (= "the way the test is written, even plain Java does not throw AME"). The point of positive test is to verify the test does what it should.

Can we also say "Did not get AbstractMethodError from JNI!" and "Did not get AbstractMethodError from Java!" in exception messages, so that it would be abundantly clear which test had failed.

**Comment by dholmes-ora** [comment]
@shipilev  Test updated as suggested.

**Review by shipilev — APPROVED** [review]
Excellent, good to go!

**Comment by dholmes-ora** [comment]
Thanks @shipilev !

/integrate
